File: crates/python/mirascript/_vm/operations/record.py
from __future__ import annotations
from typing_extensions import Sequence

from ..._helpers.types import is_vm_record
from ..._helpers.convert import to_string
from ..types import VmAny, VmRecord
from .common import AssertInit


def Omit(value: VmAny, omitted: Sequence[float | str | int]) -> VmRecord:
    AssertInit(value)
    if not is_vm_record(value):
        return {}
    result: VmRecord = {}
    valueKeys = value.keys()
    omittedSet = set([to_string(x) for x in omitted])
    for key in valueKeys:
        if key not in omittedSet:
            result[key] = value[key]
    return result


def Pick(value: VmAny, picked: Sequence[float | str | int]) -> VmRecord:
    AssertInit(value)
    if not is_vm_record(value):
        return {}
    result: VmRecord = {}
    for key in picked:
        k = to_string(key)
        if k in value:
            result[k] = value[k]

    return result

File: packages/mirascript/src/vm/operations/record.ts
import { hasOwnEnumerable, keys, setRecord } from '../../helpers/utils.js';
import { isVmRecord } from '../../helpers/types/index.js';
import type { VmAny, VmRecord, VmConst } from '../types/index.js';
import { $AssertInit } from './common.js';

/** 删除记录中的指定字段 */
export const $Omit = (value: VmAny, omitted: ReadonlyArray<number | string>): VmRecord => {
    $AssertInit(value);
    if (!isVmRecord(value)) return {};
    const result: Record<string, VmConst> = {};
    const valueKeys = keys(value);
    const omittedSet = new Set(omitted.map(String));
    for (const key of valueKeys) {
        if (!omittedSet.has(key)) {
            setRecord(result, key, value[key]);
        }
    }
    return result;
};

/** 选择记录中的指定字段 */
export const $Pick = (value: VmAny, picked: ReadonlyArray<number | string>): VmRecord => {
    $AssertInit(value);
    if (!isVmRecord(value)) return {};
    const result: Record<string, VmConst> = {};
    for (const key of picked) {
        const k = String(key);
        if (hasOwnEnumerable(value, k)) {
            setRecord(result, k, value[k]);
        }
    }
    return result;
};
